### PR TITLE
Cache specialized functions

### PR DESCRIPTION
**Pull request by cberner**
This makes queries on relatively small amounts of data that involve
parametric functions nearly 2x faster.

Fixes: https://github.com/facebook/presto/issues/2106

### CODE REVIEW

**Comment by martint** [comment]
looks good

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 1]
```diff
@@ -347,7 +364,12 @@ public FunctionInfo resolveFunction(QualifiedName name, List<TypeSignature> para
             Map<String, Type> boundTypeParameters = function.getSignature().bindTypeParameters(resolvedTypes, false, typeManager);
             if (boundTypeParameters != null) {
                 checkArgument(match == null, "Ambiguous call to %s with parameters %s", name, parameterTypes);
-                match = function.specialize(boundTypeParameters, resolvedTypes.size(), typeManager, this);
+                try {
+                    match = specializedFunctionCache.get(new SpecializedFunctionKey(function, boundTypeParameters, resolvedTypes.size()));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: If the cache doesn't throw exceptions, you should use `getUnchecked`.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 2]
```diff
@@ -347,7 +364,12 @@ public FunctionInfo resolveFunction(QualifiedName name, List<TypeSignature> para
...
+                }
+                catch (Exception e) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: Assuming you use `getUnchecked`, this should be

```
catch (ExecutionException | UncheckedExecutionException e) {
    throw Throwables.propagate(e.getCause());
```

The `getRootCause()` method digs too deep

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: For `get`, it should be 

```java
catch (ExecutionException | UncheckedExecutionException | ExecutionError e) {

**Pull request merged**